### ISSUE TITLE
Performance issues after migrating from axum-sessions

### ISSUE THREAD

**Issue by kamadorueda** [maxcountryman/tower-sessions#78]
I migrated to tower-sessions a couple of days ago, but as soon as tower-sessions landed production my server became very slow (even after multiplying the number of instances by 8x and the database capacity by 4x). Once I switched back to axum-sessions, it became fast as usual. Interestingly enough, the production servers didn't show high CPU/RAM usage, and the database was also as usual. My best guess from this behaviour is that axum-sessions uses an async RwLock (https://github.com/maxcountryman/axum-sessions/blob/799e268143a74bb03e772ae81f8391a5f68d8473/src/session.rs#L39), while tower-sessions uses a synchronous Mutex (https://github.com/maxcountryman/tower-sessions/blob/d4741d2e9d148e2c50f6aad5b02c464b3f119b70/src/session.rs#L40)

The reason why I suspect this is that when used with Axum (which is asynchronous), the synchronous Mutex blocks the event loop, preventing other asynchronous tasks from making progress. Additionally, commonly (or at least in my case), almost all my endpoints read the session, and very few write to the session, so a RwLock is much more performant as there is nothing to wait for. Plus axum-sessions had a ReadableSession and WritableSession extractor, which also boosts performance by segregating writes from reads

**Comment by maxcountryman** [comment]
Hi there, thanks for your note and sorry to hear you ran into these issues.

Were you able to do any profiling by chance? It would be helpful to see if we can positively identify where in `tower-sessions` things aren't performing. For example, I've found [`inferno`](https://crates.io/crates/inferno) helpful in the past.

We've based this design on other well-established crates, like `tower-cookies`, which [also uses a `Mutex`](https://github.com/imbolc/tower-cookies/blob/534fa426f2cfca3f50e0230fc18b18fc6f53748e/src/lib.rs#L82) to manage its inner state. (It's also worth pointing out that `tower-sessions` uses `tower-cookies` directly.)

Because the lock is only held for a brief time (as in as long as it takes to mutate the underlying hash map), workloads that don't involve a lot of write contention really shouldn't be impacted so it's a little surprising that you've run into this (again this is where profiling would provide some valuable insight to us).

That said, there's nothing really stopping us from using `RwLock` instead: nonetheless, if you're at all able to take a closer look it would be very much appreciated. 🙏

**Comment by maxcountryman** [comment]
I have a hunch that this is related to the service, since we necessarily read the inner state there which currently requires acquiring the lock.

I had put together some [prospective work to optimize this](https://github.com/maxcountryman/tower-sessions/pull/45), but wasn't able to observe any definitive performance improvements. I'm happy to resurrect that branch if you'd like to try that approach.

And of course anything you might uncover via profiling is very helpful. 😃

**Comment by furkankly** [comment]
I had the same exact issues on the free plan of fly.io and thought it was platform related, deployed it on AWS and the issues persist fyi.
I'll switch back to using this package once the issues get resolved as It has a builtin redis store impl. Thanks for it 🙏

**Comment by maxcountryman** [comment]
@furkankly are you in a position to do any profiling? It would be very helpful to see where things are bottlenecked.

Additionally I have a patch I can share which if you're feeling up to it you could try out.

**Comment by maxcountryman** [comment]
Here's a patch that uses `RwLock` instead of `Mutex`:

```diff
diff --git a/src/session.rs b/src/session.rs
index 5ead340..adec855 100644
--- a/src/session.rs
+++ b/src/session.rs
@@ -7,7 +7,7 @@ use std::{
     sync::Arc,
 };
 
-use parking_lot::Mutex;
+use parking_lot::RwLock;
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
 use serde_json::Value;
 use time::Duration;
@@ -37,7 +37,7 @@ type Data = HashMap<String, Value>;
 #[derive(Debug, Clone, Default, Serialize, Deserialize)]
 pub struct Session {
     pub(crate) id: Id,
-    inner: Arc<Mutex<Inner>>,
+    inner: Arc<RwLock<Inner>>,
 }
 
 impl Session {
@@ -70,7 +70,7 @@ impl Session {
 
         Self {
             id: Id::default(),
-            inner: Arc::new(Mutex::new(inner)),
+            inner: Arc::new(RwLock::new(inner)),
         }
     }
 
@@ -117,7 +117,7 @@ impl Session {
     /// assert_eq!(value, Some(serde_json::json!(42)));
     /// ```
     pub fn insert_value(&self, key: &str, value: Value) -> Option<Value> {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         if inner.data.get(key) != Some(&value) {
             inner.modified_at = Some(OffsetDateTime::now_utc());
             inner.data.insert(key.to_string(), value)
@@ -162,7 +162,7 @@ impl Session {
     /// assert_eq!(value, serde_json::json!(42));
     /// ```
     pub fn get_value(&self, key: &str) -> Option<Value> {
-        let inner = self.inner.lock();
+        let inner = self.inner.read();
         inner.data.get(key).cloned()
     }
 
@@ -211,7 +211,7 @@ impl Session {
     /// assert!(value.is_none());
     /// ```
     pub fn remove_value(&self, key: &str) -> Option<Value> {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         if let Some(removed) = inner.data.remove(key) {
             inner.modified_at = Some(OffsetDateTime::now_utc());
             Some(removed)
@@ -257,7 +257,7 @@ impl Session {
         old_value: impl Serialize,
         new_value: impl Serialize,
     ) -> Result<bool> {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         match inner.data.get(key) {
             Some(current_value) if serde_json::to_value(&old_value)? == *current_value => {
                 let new_value = serde_json::to_value(&new_value)?;
@@ -284,7 +284,7 @@ impl Session {
     /// assert!(session.get_value("foo").is_none());
     /// ```
     pub fn clear(&self) {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         inner.data.clear();
     }
 
@@ -306,17 +306,17 @@ impl Session {
     /// assert!(matches!(session.deleted(), Some(Deletion::Deleted)));
     /// ```
     pub fn delete(&self) {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         inner.deleted = Some(Deletion::Deleted);
     }
 
     pub(crate) fn reset_deleted(&self) {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         inner.deleted = None;
     }
 
     pub(crate) fn reset_modified(&self) {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         inner.modified_at = None;
     }
 
@@ -342,7 +342,7 @@ impl Session {
     /// ));
     /// ```
     pub fn cycle_id(&self) {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         inner.deleted = Some(Deletion::Cycled(self.id));
         inner.modified_at = Some(OffsetDateTime::now_utc());
     }
@@ -400,7 +400,7 @@ impl Session {
     /// );
     /// ```
     pub fn expiry(&self) -> Option<Expiry> {
-        let inner = self.inner.lock();
+        let inner = self.inner.read();
         inner.expiry.clone()
     }
 
@@ -426,7 +426,7 @@ impl Session {
     /// assert!(session.is_modified());
     /// ```
     pub fn set_expiry(&self, expiry: Option<Expiry>) {
-        let mut inner = self.inner.lock();
+        let mut inner = self.inner.write();
         inner.expiry = expiry;
         inner.modified_at = Some(OffsetDateTime::now_utc());
     }
@@ -443,7 +443,7 @@ impl Session {
     /// assert!(session.expiry_date() > OffsetDateTime::now_utc());
     /// ```
     pub fn expiry_date(&self) -> OffsetDateTime {
-        let inner = self.inner.lock();
+        let inner = self.inner.read();
         match inner.expiry {
             Some(Expiry::OnInactivity(duration)) => {
                 let modified_at = inner.modified_at.unwrap_or_else(OffsetDateTime::now_utc);
@@ -493,7 +493,7 @@ impl Session {
     /// assert!(session.is_modified());
     /// ```
     pub fn is_modified(&self) -> bool {
-        let inner = self.inner.lock();
+        let inner = self.inner.write();
         inner.modified_at.is_some() && !inner.data.is_empty()
     }
 
@@ -521,7 +521,7 @@ impl Session {
             return Some(Deletion::Deleted);
         };
 
-        self.inner.lock().deleted
+        self.inner.read().deleted
     }
 
     /// Returns `true` if the session is empty.
@@ -538,7 +538,7 @@ impl Session {
     /// assert!(!session.is_empty());
     /// ```
     pub fn is_empty(&self) -> bool {
-        self.inner.lock().data.is_empty()
+        self.inner.read().data.is_empty()
     }
 }
```

In a simple load test, I can't seem to observe a difference. However, that isn't a good proxy for a real application's workload, so it's possible this does help. That said, the ideal case here is that we have some direction from profiling instrumentation (even if this does seem to help).

**Comment by abcd-ca** [comment]
I'm not sure what kind of profiling you're looking for but I tried using [drill](https://github.com/fcsonline/drill) with this benchmark.yml file:

```yaml
---

concurrency: 500
base: 'http://localhost:8080/api'
iterations: 10000
rampup: 2

plan:
  - name: Health Check
    request:
      url: /health_check
```

I am running the load balancer like this: `drill --benchmark benchmark.yml --stats --no-check-certificate`

With a `MemoryStore`, all the requests succeed. With a `RedisStore`, my REST API Rust App hangs after very few requests – I counted half a dozen once. No panic from my app, just hangs so I suspect a deadlock. Note that my `health_check` route doesn't use the session directly but I guess Axum provides it to every route because of the line I have when configuring the Axum router: 

```rust
.layer(session_service)
```

**Comment by maxcountryman** [comment]
If you're using Redis via our store then it's important to set a reconnection policy. If you don't the connection will die. This manifests exactly as you've described.

**Comment by abcd-ca** [comment]
Thanks @maxcountryman. I'm just getting to know Redis. Is this what you mean?
```rust
let reconnect_policy = ReconnectPolicy::Constant {
        attempts: 1,      // initial attempts
        max_attempts: 10, // max attempts
        delay: 5_000,     // delay in milliseconds
        jitter: 100,      // jitter in milliseconds
    };

    let redis_client = RedisClient::new(
        redis_config,
        None,
        None,
        Some(reconnect_policy),
    );
```

**Comment by maxcountryman** [comment]
@abcd-ca can we move your Redis-related questions into a separate [Discussions](https://github.com/maxcountryman/tower-sessions/discussions) thread? I suspect this will also be helpful for other folks and I do want to make sure we get to the bottom of other performance issues folks are running into here so it'll help if we can focus the issue thread. Happy to help over there. 🙂

**Comment by maxcountryman** [comment]
It's possible the overhead of deleting unused sessions is a contributing factor here. I've opened https://github.com/maxcountryman/tower-sessions/pull/90 to address.

**Comment by furkankly** [comment]
> If you're using Redis via our store then it's important to set a reconnection policy. If you don't the connection will die. This manifests exactly as you've described.

I didn't have enough time to figure out how to do the profiling yet but since I am also using the built-in RedisStore and having the same problem (the server getting very slow with no panics), Ive put together a new RedisStore changing the existing one which uses a connection pool instead of a single client:
([ref1](https://github.com/aembke/fred.rs/blob/76d2be222bcc1b33b028cf0d911a159d814430c1/examples/pool.rs), [ref2](https://crates.io/crates/async-fred-session/0.1.5))

```
use async_trait::async_trait;
use time::OffsetDateTime;
use tower_sessions::fred::{
    prelude::{KeysInterface, RedisPool},
    types::Expiration,
};
use tower_sessions::{session::Id, Session, SessionStore};

/// An error type for `RedisStore`.
#[derive(thiserror::Error, Debug)]
pub enum RedisStoreError {
    /// A variant to map to `fred::error::RedisError` errors.
    #[error("Redis error: {0}")]
    Redis(#[from] tower_sessions::fred::error::RedisError),

    /// A variant to map `rmp_serde` encode errors.
    #[error("Rust MsgPack encode error: {0}")]
    RmpSerdeEncode(#[from] rmp_serde::encode::Error),

    /// A variant to map `rmp_serde` decode errors.
    #[error("Rust MsgPack decode error: {0}")]
    RmpSerdeDecode(#[from] rmp_serde::decode::Error),
}

/// A Redis session store.
#[derive(Debug, Clone)]
pub struct RedisStore {
    pool: RedisPool,
}

impl RedisStore {
    pub fn new(pool: RedisPool) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl SessionStore for RedisStore {
    type Error = RedisStoreError;

    async fn save(&self, session: &Session) -> Result<(), Self::Error> {
        let expire = Some(session.expiry_date())
            .map(OffsetDateTime::unix_timestamp)
            .map(Expiration::EXAT);

        self.pool
            .set(
                session.id().to_string(),
                rmp_serde::to_vec(&session)?.as_slice(),
                expire,
                None,
                false,
            )
            .await?;

        Ok(())
    }

    async fn load(&self, session_id: &Id) -> Result<Option<Session>, Self::Error> {
        let data = self
            .pool
            .get::<Option<Vec<u8>>, _>(session_id.to_string())
            .await?;

        if let Some(data) = data {
            Ok(Some(rmp_serde::from_slice(&data)?))
        } else {
            Ok(None)
        }
    }

    async fn delete(&self, session_id: &Id) -> Result<(), Self::Error> {
        self.pool.del(session_id.to_string()).await?;
        Ok(())
    }
}
```
and using this store with a reconnection policy set on the call site as:
```
    let redis_url = env::var("REDIS_URL").expect("REDIS_URL env var not found");
    let redis_config = RedisConfig::from_url(redis_url.as_str()).unwrap();
    let redis_pool = Builder::from_config(redis_config)
        .set_policy(ReconnectPolicy::new_exponential(3, 100, 30_000, 2))
        .build_pool(5)
        .unwrap();
    redis_pool.connect();
    redis_pool
        .wait_for_connect()
        .await
        .expect("Couldn't connect to redis sessions store");
    let session_store = RedisStore::new(redis_pool);
```
but the same issue persists. And thats why Im sharing this here, happy to move this to Redis discussion as well if you want. @maxcountryman 

Its probably the event loop getting blocked somewhere as OP pointed out. Its a side project with no users and it only takes to send a few requests for the server to start hanging with no actual resource usage. Whats interesting is that, this happens only on production for me. Im using the same Upstash server with fly and ec2 and when I use axum-sessions it works, when I switch to this package, it starts hanging after a few requests.

**Comment by maxcountryman** [comment]
> Its probably the event loop getting blocked somewhere as OP pointed out.

It's very difficult to say without some kind of profiling visibility. Note [tokio's recommendations](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use) re synchronous lock. I don't think we're following a bad pattern here.

I've also noticed a problem with `fred` previously, but I was able to address that with a reconnection policy which is why I mentioned it.

**Comment by maxcountryman** [comment]
@furkankly I'm not sure if it'll make any difference with your specific issue (since again, it sounds like we're running into a problem that's related to `fred` itself), but you could try https://github.com/maxcountryman/tower-sessions/pull/91 and see if removing support for concurrent data access via shared memory helps you.

**Comment by heat1q** [comment]
I don't think this issue is only related to redis. I noticed perfomance bottlenecks and even deadlocks using either redis (both with `fred` and `redis` clients) and postgres if the load increases. 

I was actually able to reproduce this with the examples (`redis-store`, `postgres-store`). If I also [restrict](https://docs.rs/tokio/latest/tokio/attr.main.html#current-thread-runtime) the tokio runtime to a single thread (simulating a single-core machine), I ran into a deadlock with only 2 concurrent requests using any load testing tool.

I'm not really sure where the issue is here considering the `SessionStore` is just a simple wrapper around the upstream client(s). 

The `MemoryStore` seems to work without any issues though.

**Comment by maxcountryman** [comment]
Please include a reproduction so we can take a closer look.

**Comment by maxcountryman** [comment]
@heat1q I am not able to reproduce this with `current_thread` and the`redis-store` example on bb06fd2619b8c4416186f08ad0503f74b0cb0243.

For example, here's a sample output: https://gist.github.com/maxcountryman/898d86930a677a0fde4ae34d03c01f64

Folks, if you're going to comment here then please provide:

1. Profiling data, using inferno or similar tools.
2. A complete application, which we can run, that demonstrates the issue.

Without these things, it's very difficult to provide help and moreover we are all just speculating as to why things aren't working as expected.

**Comment by heat1q** [comment]
You are absolutely right. I initially thought it was easily reproducible given my description. However, I missed an important detail: You have to include a valid session id in the cookie so that each request triggers a `load` in `SessionStore`.

Let me summarize the steps needed to reproduce, demonstrated with the `redis-store` example:

1. using a single threaded runtime 
   ```diff
   diff --git a/examples/redis-store.rs b/examples/redis-store.rs
   index a9e9003..9f10f40 100644
   --- a/examples/redis-store.rs
   +++ b/examples/redis-store.rs
   @@ -14,7 +14,7 @@ const COUNTER_KEY: &str = "counter";
    #[derive(Serialize, Deserialize, Default)]
    struct Counter(usize);

   -#[tokio::main]
   +#[tokio::main(flavor = "current_thread")]
    async fn main() -> Result<(), Box<dyn std::error::Error>> {
        let client = RedisClient::default();
   ```
2. perform a "manual" call and get a valid active session id from the `set-cookie` header
   ```
   curl -v --silent localhost:3000 --stderr - | grep set-cookie
   ```
3. use that session id and perform the load test with 2 or more concurrent requests. (I'm using the minimum number of concurrent request to reduce noise, but any number `>= 2` should do the trick)
   ```yaml
   concurrency: 2
   base: "http://localhost:3000"
   iterations: 2

   plan:
     - name: Index
        request:
          url: /
          headers:
            Cookie: id=<session-id>
   ```
 4. the server is now unresponsive. See the [output](https://gist.github.com/heat1q/c1ea877a60305de1727cbe3d5bd855bf) of `drill`.

I can provide profiling data if you wish. However, I would first like to confirm this issue before going further into the specifics as to why and where this happens.

**Comment by maxcountryman** [comment]
@heat1q Thank you for that.

I tried 24514545e5b578485c94fc679dd61ac980f07913 and it /appears/ to be fixed in that commit. Can you confirm you see the same?

**Comment by heat1q** [comment]
@maxcountryman Awesome, this indeed resolves the issue! I also tested it in my end application and it works flawless with `v0.6.0`; no bottlenecks or unresponsiveness. 

Maybe @kamadorueda can confirm this fix as well, then we can close this issue?

**Comment by furkankly** [comment]
I can confirm that the issue I had is fixed with `v0.6.0` as well. Thank you folks! 🎉

**Comment by maxcountryman** [comment]
Thanks for verifying! I'll close this for now. If folks encounter new issues, please open a discussion so we can investigate.
